domain/schema/model.go

```diff
@@ -401,23 +402,246 @@ ON charm_storage (charm_uuid);
+    CONSTRAINT       fk_machine_filesystem_machine
+        FOREIGN KEY      (machine_uuid)
+        REFERENCES       machine(uuid),
+    CONSTRAINT       fk_machine_filesystem_filesystem
+        FOREIGN KEY      (filesystem_uuid)
+        REFERENCES       storage_filesystem(uuid),
+    PRIMARY KEY (machine_uuid, filesystem_uuid)
+);
+
+CREATE TABLE address_type (
+    id INT PRIMARY KEY,
+    value TEXT NOT NULL
+);
+
+INSERT INTO address_type VALUES
+    (0, 'hostname'),
+    (1, 'ipv4'),
+    (2, 'ipv6');
+
+CREATE TABLE address_scope (
+    id INT PRIMARY KEY,
+    value TEXT NOT NULL
+);
+
+INSERT INTO address_scope VALUES
+    (0, 'unknown'),
+    (1, 'public'),
+    (2, 'local-cloud'),
+    (3, 'local-machine'),
+    (4, 'link-local');
+
+CREATE TABLE address_origin (
+    id INT PRIMARY KEY,
+    value TEXT NOT NULL
+);
+
+INSERT INTO address_origin VALUES
+    (0, 'unknown'),
+    (1, 'provider'),
+    (2, 'machine');
+
+CREATE TABLE address (
     uuid            TEXT PRIMARY KEY,
-    machine_id      TEXT NOT NULL,
-    net_node_uuid   TEXT NOT NULL,
-    life_id         INT NOT NULL,
-    CONSTRAINT      fk_machine_net_node
-        FOREIGN KEY (net_node_uuid)
-        REFERENCES  net_node(uuid),
-    CONSTRAINT      fk_machine_life
-        FOREIGN KEY (life_id)
-        REFERENCES  life(id)
+    value           TEXT NOT NULL,
+    address_type_id TEXT NOT NULL,
+    scope_id        TEXT NOT NULL,
+    origin_id       TEXT NOT NULL,
+    space_id        TEXT NOT NULL,
+    CONSTRAINT      fk_address_address_type
+        FOREIGN KEY     (address_type_id)
+        REFERENCES      address_type(id),
+    CONSTRAINT      fk_address_space
+        FOREIGN KEY     (scope_id)
+        REFERENCES      address_scope(id),
+    CONSTRAINT      fk_address_origin
+        FOREIGN KEY     (origin_id)
+        REFERENCES      address_origin(id),
+    CONSTRAINT      fk_address_space
+        FOREIGN KEY     (space_id)
+        REFERENCES      space(uuid)
+);
+
+CREATE INDEX idx_address_address
+ON address (uuid);
+
+CREATE TABLE machine_instance_address (
+    machine_uuid     TEXT NOT NULL,
+    address_uuid     TEXT NOT NULL,
+    CONSTRAINT       fk_machine_address_machine
+        FOREIGN KEY      (machine_uuid)
+        REFERENCES       machine(uuid),
+    CONSTRAINT       fk_machine_address_address
+        FOREIGN KEY      (address_uuid)
+        REFERENCES       address(uuid),
+    PRIMARY KEY (machine_uuid, address_uuid)
+);
+
+CREATE TABLE machine_machine_address (
+    machine_uuid            TEXT NOT NULL,
+    machine_address_uuid    TEXT NOT NULL,
+    CONSTRAINT              fk_machine_address_machine
+        FOREIGN KEY             (machine_uuid)
+        REFERENCES              machine(uuid),
+    CONSTRAINT              fk_machine_address_address
+        FOREIGN KEY             (machine_address_uuid)
+        REFERENCES              address(uuid),
+    PRIMARY KEY (machine_uuid, machine_address_uuid)
```

Comment:
What was the difference between the instance and machine addresses? Is it the "desired" vs "assigned" addresses? In that case why do we need the join table `machine_machine_address`?

Reply:
No AIUI the instance address is the addresses for the machine on the cloud network (instancepoller hits the provider to get this), and the machine addresses is the addresses local to the machine obtained from the machine itself (which is related to the local interfaces like NICs bridges etc).